Source/render.cpp

```diff
@@ -139,48 +139,56 @@ inline static void RenderLine(BYTE **dst, BYTE **src, int n, BYTE *tbl, DWORD ma
 
 #ifdef NO_OVERDRAW
 	if (*dst < gpBufStart || *dst > gpBufEnd) {
-		*src += n;
-		*dst += n;
-		return;
+		goto skip;
 	}
 #endif
 
 	if (mask == 0xFFFFFFFF) {
 		if (light_table_index == lightmax) {
 			memset(*dst, 0, n);
-			(*src) += n;
-			(*dst) += n;
 		} else if (light_table_index == 0) {
 			memcpy(*dst, *src, n);
-			(*src) += n;
-			(*dst) += n;
 		} else {
-			for (i = 0; i < n; i++, (*src)++, (*dst)++) {
-				(*dst)[0] = tbl[(*src)[0]];
+			for (i = 0; i < n; i++) {
+				(*dst)[i] = tbl[(*src)[i]];
 			}
 		}
 	} else {
+		// The number of iterations is anyway limited by the size of the mask.
+		// So we can limit it by ANDing the mask with another mask that only keeps
+
+		// iterations that are lower than n. We can now avoid testing if i < n
+		// at every loop iteration.
+		mask &= ((((DWORD)1) << n) - 1) << ((sizeof(DWORD) * CHAR_BIT) - n);
+
+#if defined(__GNUC__) || defined(__clang__)
+#define SKIP_ZERO_BITS(i, mask) { int z = __builtin_clz(mask); i += z, mask <<= z; }
+#else
+#define SKIP_ZERO_BITS(i, mask) for (; (mask & 0x80000000) == 0; i++, mask <<= 1);
+#endif
+
+#define FOREACH_SET_BIT(i, mask, ...) \

```

Comment:
Not a big fan of preprocessor code when it's not strictly needed

Reply:
I can write it with lambdas if required

Reply:
lambda sounds like the better option